### PR TITLE
Reduce size_of<HashMap> by one pointer

### PR DESCRIPTION
**Pull request by SimonSapin**
Fixes https://github.com/rust-lang/hashbrown/issues/69 to the extent possible without more dramatic changes.

### CODE REVIEW

**Comment by SimonSapin** [comment]
`cargo bench` before:

```rust
test insert_erase_i32 ... bench:      15,441 ns/iter (+/- 47)
test insert_erase_i64 ... bench:      14,978 ns/iter (+/- 72)
test insert_i32       ... bench:      11,097 ns/iter (+/- 63)
test insert_i64       ... bench:      10,759 ns/iter (+/- 56)
test iter_i32         ... bench:       1,529 ns/iter (+/- 2)
test iter_i64         ... bench:       1,530 ns/iter (+/- 7)
test lookup_fail_i32  ... bench:       1,798 ns/iter (+/- 0)
test lookup_fail_i64  ... bench:       1,798 ns/iter (+/- 2)
test lookup_i32       ... bench:       2,182 ns/iter (+/- 1)
test lookup_i64       ... bench:       2,214 ns/iter (+/- 52)
```

After: 

```rust
test insert_erase_i32 ... bench:      17,152 ns/iter (+/- 25)
test insert_erase_i64 ... bench:      24,686 ns/iter (+/- 15)
test insert_i32       ... bench:      12,157 ns/iter (+/- 7)
test insert_i64       ... bench:      19,291 ns/iter (+/- 14)
test iter_i32         ... bench:       1,529 ns/iter (+/- 0)
test iter_i64         ... bench:       1,534 ns/iter (+/- 1)
test lookup_fail_i32  ... bench:       2,609 ns/iter (+/- 0)
test lookup_fail_i64  ... bench:       2,826 ns/iter (+/- 0)
test lookup_i32       ... bench:       2,609 ns/iter (+/- 1)
test lookup_i64       ... bench:       2,807 ns/iter (+/- 331)
```

The run time cost appears quite significant, I don’t know if it can be improved much.

File: src/raw/mod.rs [hunk 1]
```diff
@@ -422,13 +416,19 @@ impl<T> RawTable<T> {
         dealloc(self.ctrl.as_ptr(), layout);
     }
 
+    // Pointer to the array of buckets
+    fn data(&self) -> NonNull<T> {
+        let (_, data_offset) = calculate_layout::<T>(self.buckets()).unwrap();
```
Comment (stjepang) [inline src/raw/mod.rs]: Can you try replacing `unwrap()` with `unwrap_or_else(|| hint::unreachable_unchecked())`?

**Comment by alkis** [comment]
`data` and `ctrl` are hot fields, you likely do not want to access them through arithmetic. How about reducing size by putting `growth_left` into the buckets array. `growth_left` is only accessed when inserting and that's expensive by itself.

File: src/raw/mod.rs (hunk 1 above)
Comment (Amanieu) [reply src/raw/mod.rs]: Try calculating `data_offset` manually by copying the code from `calculate_layout`.

Comment (Amanieu) [reply src/raw/mod.rs]: Also you can make this function return a `*mut T` instead of a `NonNull<T>`.

**Comment by SimonSapin** [comment]
```diff
diff --git a/src/raw/mod.rs b/src/raw/mod.rs
index 7d18bc0..5d58074 100644
--- a/src/raw/mod.rs
+++ b/src/raw/mod.rs
@@ -418,7 +418,13 @@ impl<T> RawTable<T> {
 
     // Pointer to the array of buckets
     fn data(&self) -> NonNull<T> {
-        let (_, data_offset) = calculate_layout::<T>(self.buckets()).unwrap();
+        let buckets = self.buckets();
+
+        // Inline parts of `calculate_layout()`
+        debug_assert!(buckets.is_power_of_two());
+        let data_align = usize::max(mem::align_of::<T>(), Group::WIDTH);
+        let data_offset = (buckets + Group::WIDTH + data_align - 1) & !(data_align - 1);
+
         unsafe { NonNull::new_unchecked(self.ctrl.as_ptr().add(data_offset) as *mut T) }
     }
 
```

Compared to this PR, also adding the diff above slightly improves some benchmarks but make others much worse.

```rust
test insert_erase_i32 ... bench:      22,672 ns/iter (+/- 42)
test insert_erase_i64 ... bench:      23,299 ns/iter (+/- 101)
test insert_i32       ... bench:      18,253 ns/iter (+/- 389)
test insert_i64       ... bench:      18,440 ns/iter (+/- 1,060)
test iter_i32         ... bench:       1,537 ns/iter (+/- 2)
test iter_i64         ... bench:       1,531 ns/iter (+/- 1)
test lookup_fail_i32  ... bench:       1,635 ns/iter (+/- 0)
test lookup_fail_i64  ... bench:       1,581 ns/iter (+/- 0)
test lookup_i32       ... bench:       2,477 ns/iter (+/- 7)
test lookup_i64       ... bench:       2,520 ns/iter (+/- 9)
```

Same hardware as https://github.com/rust-lang/hashbrown/pull/73#issuecomment-487050061

----

Closing since this particular approach is likely to always have significant perf cost. Anyone feel free to make other PRs to try other ideas.

**Pull request closed without merging**